### PR TITLE
[LLDB] Support exception breakpoints for plugin-provided languages

### PR DESCRIPTION
**Pull request by walter-erquinigo**
CommandObjectBreakpoint has a harcoded list of languages for which exception breakpoints can be enabled. I'm making this a bit more generic so that my Mojo plugin can get this feature.
Basically, I'm adding a new overridable method `Language::SupportsExceptionBreakpoints` that can be used by language plugins to determine whether they support this feature or not. This method is used in addition to the hardcoded list and, as an example, I'm using it for the ObjC language support.

Another route is simply to avoid doing the check that it's being done right now and simply try to the create the exception breakpoint for whatever language that is not in the hardcoded list. I'm happy to do that if the reviewers think it's a good idea.

As a note, the other possible place for adding this `SupportsExceptionBreakpoints` method is in `LanguageRuntime`. However, accessing it requires having a process, which is not always the case when invoking the `breakpoint set -E` command. The process might not be alive yet, so `Language` is a good second place for this.

And as a final note, I don't want to make this `SupportsExceptionBreakpoints` complicated. I'm keeping it as simple as possible because it can easily evolve as it's not part of the public API.

### CODE REVIEW

**Comment by DavidSpickett** [comment]
> This method is used in addition to the hardcoded list and, as an example, I'm using it for the ObjC language support.

And does this list include languages that are not plugin provided?

I was wondering why only Objective C changed, but if the rest are not plugins then this makes sense.

File: lldb/include/lldb/Target/Language.h [hunk 1]
```diff
@@ -245,7 +245,7 @@ class Language : public PluginInterface {
   // a match.  But we wouldn't want this to match AnotherA::my_function.  The
   // user is specifying a truncated path, not a truncated set of characters.
   // This function does a language-aware comparison for those purposes.
-  virtual bool DemangledNameContainsPath(llvm::StringRef path, 
+  virtual bool DemangledNameContainsPath(llvm::StringRef path,
```
Comment (DavidSpickett) [inline lldb/include/lldb/Target/Language.h:248]: Unrelated change.

**Review by DavidSpickett — APPROVED** [review]
LGTM

File: lldb/source/Commands/CommandObjectBreakpoint.cpp [hunk 2]
```diff
@@ -319,6 +316,12 @@ class CommandObjectBreakpointSet : public CommandObjectParsed {
           error_context = "Unknown language type for exception breakpoint";
           break;
         default:
+          if (Language *languagePlugin = Language::FindPlugin(language)) {
```
Comment (DavidSpickett) [inline lldb/source/Commands/CommandObjectBreakpoint.cpp:319]: You could drop the braces from the first if I think, but this is already in so many sets of braces I'd keep them for clarity's sake.

Comment (walter-erquinigo) [reply lldb/source/Commands/CommandObjectBreakpoint.cpp:319]: Within switch statements I tend to be very paranoid with braces

**Comment by walter-erquinigo** [comment]
> And does this list include languages that are not plugin provided?

The hardcoded list in the file being changes includes C++ and Objc in its various variants.

> I was wondering why only Objective C changed, but if the rest are not plugins then this makes sense.

That was the only simple example I could use to show how this change does work. 

Thanks for the review!

**Comment by jimingham** [comment]
> CommandObjectBreakpoint has a harcoded list of languages for which exception breakpoints can be enabled. I'm making this a bit more generic so that my Mojo plugin can get this feature. Basically, I'm adding a new overridable method `Language::SupportsExceptionBreakpoints` that can be used by language plugins to determine whether they support this feature or not. This method is used in addition to the hardcoded list and, as an example, I'm using it for the ObjC language support.
> 
> Another route is simply to avoid doing the check that it's being done right now and simply try to the create the exception breakpoint for whatever language that is not in the hardcoded list. I'm happy to do that if the reviewers think it's a good idea.
> 
> As a note, the other possible place for adding this `SupportsExceptionBreakpoints` method is in `LanguageRuntime`. However, accessing it requires having a process, which is not always the case when invoking the `breakpoint set -E` command. The process might not be alive yet, so `Language` is a good second place for this.
> 

Exception Breakpoints are created by the static method LanguageRuntime::CreateExceptionBreakpoint.  So they don't require a process.  Exception breakpoint resolvers can be "resolver discovering resolvers".  This used to be necessary when there were two ObjC runtimes possible on MacOS X, and you couldn't necessarily tell which one was going to be used till the program till it ran.  So they are able to adjust themselves when their Target launches a process.

It's a little awkward to have a static method in LanguageRuntime that makes the Exception breakpoint conditioned by a Language method 'SupportsExceptionBreakpoint".

> And as a final note, I don't want to make this `SupportsExceptionBreakpoints` complicated. I'm keeping it as simple as possible because it can easily evolve as it's not part of the public API.

**Comment by walter-erquinigo** [comment]
> Exception Breakpoints are created by the static method LanguageRuntime::CreateExceptionBreakpoint. So they don't require a process. 

Indeed, but let me elaborate. What I need is a method that doesn't require a process that could tell me if a certain language can support exception breakpoints, as that's what the checks in the `breakpoint set -E <lang>` command try to do.

Then I had two main choices w.r.t `LanguageRuntime`:
- Add the method as non-static `LanguageRuntime`, but `LanguageRuntime` instances do require a process.
- Add a static method in `LanguageRuntime`, like `CreateExceptionBreakpoint`, but it turns out that `CreateExceptionBreakpoint` ends up relying on `Language::GetDefaultExceptionResolverDescription` and `Language::GetExceptionResolverDescription`. So this made me believe that it'd be better to place all the static exception feature bits in `Language` and not `LanguageRuntime`. `LanguageRuntime` seems to be a place more suited for dealing with runtime behaviors and not static pieces of information.

> It's a little awkward to have a static method in LanguageRuntime that makes the Exception breakpoint conditioned by a Language method 'SupportsExceptionBreakpoint".

I somewhat agree with it, but probably I'd end up writing a static method `LanguageRuntime::SupportsExceptionBreakpoint` that invokes a non-static `Language::SupportsExceptionBreakpoint`. And as I mentioned, there's already precedent for other exception related stuff in `Language`.
